### ISSUE TITLE
Collision of Systematics

### ISSUE THREAD

**Issue by johnda102** [UCATLAS/xAODAnaHelpers#293]
HI All, 

  I have a job setup where I do:
   -) muon calibration (without Muon systematics for the moment) 
  -) then JetCalibration with systematics ("All", 1).

Im seeing the following jet systematics configured:

JetCalibrator::initial... INFO     Initializing JER Systematics :
JetCalibrator::initial... INFO    JERSysList Empty at index 0.
JetCalibrator::initial... INFO  
JetCalibrator::initial... INFO       MUONS_ID__1down
JetCalibrator::initial... INFO       MUONS_ID__1up
JetCalibrator::initial... INFO       MUONS_MS__1down
JetCalibrator::initial... INFO       MUONS_MS__1up
JetCalibrator::initial... INFO       MUONS_SCALE__1down
JetCalibrator::initial... INFO       MUONS_SCALE__1up
JetCalibrator::initial... INFO       JET_GroupedNP_1__1up
JetCalibrator::initial... INFO       JET_GroupedNP_1__1down
JetCalibrator::initial... INFO       JET_GroupedNP_2__1up
JetCalibrator::initial... INFO       JET_GroupedNP_2__1down
JetCalibrator::initial... INFO       JET_GroupedNP_3__1up
JetCalibrator::initial... INFO       JET_GroupedNP_3__1down
JetCalibrator::initial... INFO       JET_JER_SINGLE_NP__1up

Im making systematically varried jet collections for the muon systematics. 

Im not 100% sure whats gong on, b/c I think there is some magic going on behind the scenes in the following calls.

``` c++
const CP::SystematicRegistry& systReg = CP::SystematicRegistry::getInstance();
 const CP::SystematicSet& recSyst = (systReg.recommendedSystematics());
```

I think what is happening is that Im initializing the muon cp tools which talks to the SystematicRegistry even though im not specifying to use muons systematics.  These muon systematics then get picked up in the JetCalibration when we make similar calls to the systematicRegistry. 

Does anyone have a better understanding of what is going on here ?  How do we avoid this ?

**Comment by mmilesi** [comment]
Hi @johnda102 ,

interesting, I never saw any of this. Can you send around the exact configuration you are using?

**Comment by johnda102** [comment]
The configuration is here:

https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/JDM/hh4b/Run2/Code/XhhCommon/trunk/scripts/combinedMaster.py

which sets up the BasicEventSelection and the MuonCalibrator (which is commented out until we fix this issue) and then calls the resolved selection here:

https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/JDM/hh4b/Run2/Code/XhhCommon/trunk/scripts/XhhResolved_config.py

which sets up the JetCalbration.

**Comment by gfacini** [comment]
Hi John,

Are you actually seeing those variations run?

Here the list of recommended systematics is
pulled from the JES tool itself:
https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L366

then used here to get the list of systs to run over
https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L378

Cheers,
Gabriel

**Comment by johnda102** [comment]
HI Gabriel,

  Yes when the MuonCalibrator is configured. I see the following jet
variations run:

JetCalibrator::initial... INFO MUONS_ID__1down
JetCalibrator::initial... INFO MUONS_ID__1up
JetCalibrator::initial... INFO MUONS_MS__1down
JetCalibrator::initial... INFO MUONS_MS__1up
JetCalibrator::initial... INFO MUONS_SCALE__1down
JetCalibrator::initial... INFO MUONS_SCALE__1up
JetCalibrator::initial... INFO JET_GroupedNP_1__1up
JetCalibrator::initial... INFO JET_GroupedNP_1__1down
JetCalibrator::initial... INFO JET_GroupedNP_2__1up
JetCalibrator::initial... INFO JET_GroupedNP_2__1down
JetCalibrator::initial... INFO JET_GroupedNP_3__1up
JetCalibrator::initial... INFO JET_GroupedNP_3__1down
JetCalibrator::initial... INFO JET_JER_SINGLE_NP__1up

When the muon calibrator is not configured I only see the JET_* variations.

The presence of the muon calbration affects the line
https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L341

the variations you pointed to in line:
https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L366

get added to the ones returned earlier, which arleady have the MUON_*
variations when the muoncalibrator is configed.



On Mon, Oct 19, 2015 at 9:25 AM, gfacini <notifications@github.com> wrote:

> Hi John,
>
> Are you actually seeing those variations run?
>
> Here the list of recommended systematics is
> pulled from the JES tool itself:
>
> https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L366
>
> then used here to get the list of systs to run over
>
> https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L378
>
> Cheers,
> Gabriel
>
>
>
> On Sun, Oct 18, 2015 at 2:55 PM, johnda102 <notifications@github.com>
> wrote:
>
> > The configuration is here:
> >
> >
> >
> https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/JDM/hh4b/Run2/Code/XhhCommon/trunk/scripts/combinedMaster.py
> >
> > which sets up the BasicEventSelection and the MuonCalibrator (which is
> > commented out until we fix this issue) and then calls the resolved
> > selection here:
> >
> >
> >
> https://svnweb.cern.ch/trac/atlasphys-exo/browser/Physics/Exotic/JDM/hh4b/Run2/Code/XhhCommon/trunk/scripts/XhhResolved_config.py
> >
> > which sets up the JetCalbration.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/UCATLAS/xAODAnaHelpers/issues/293#issuecomment-149008805
> >
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/UCATLAS/xAODAnaHelpers/issues/293#issuecomment-149126162>
> .
>



-- 
University of Chicago
http://hep.uchicago.edu/~johnda

**Comment by gfacini** [comment]
Hi John,


Super confused at the moment.

The full set of systematics is added here to m_systList

https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L347

then each JES error is added *again*? here

https://github.com/UCATLAS/xAODAnaHelpers/blob/master/Root/JetCalibrator.cxx#L384

...?

Why do we need line 347 anyway?

Gabriel

**Comment by johnda102** [comment]
So thinking about this more...

Is there any reason to use the CP::SystematicRegistry in the calibrators  ?

eg:
const CP::SystematicRegistry& systReg = CP::SystematicRegistry::getInstance();                                                                    
const CP::SystematicSet& recSyst = (systReg.recommendedSystematics());    

As opposed to just getting the systematics directly from the CP tools themselves ?

Avoiding the SystematicRegistry would solve the systematic clash and it is not needed in the JetCalibrator. 

My guess is that this same issue would come up with combining other calibrators, and that we should just remove the SystematicRegistry  calls everywhere, but maybe there is something im missing...

**Comment by mmilesi** [comment]
Hi @johnda102 ,

I will check what you did and then change the electron and muon calibrators accordingly

Marco

**Comment by jdandoy** [comment]
This was addressed by #297 as being fixed, so I'll close it.
